akka-stream/src/main/scala/akka/stream/impl/ActorRefBackpressureSinkActor.scala

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import java.util
+
+import akka.actor._
+import akka.dispatch.sysmsg.{ DeathWatchNotification, SystemMessage, Watch }
+import akka.stream.ActorMaterializer
+import akka.stream.Attributes.InputBuffer
+import akka.stream.actor.RequestStrategy
+import akka.stream.stage._
+
+import scala.concurrent.Future
+
+/**
+ * INTERNAL API
+ */
+private[akka] class ActorRefBackpressureSinkStage[Out](ref: ActorRef, onInitMessage: Any,
+                                                       ackMessage: Any,
+                                                       onCompleteMessage: Any)
+  extends SinkStage[Out, Unit]("ActorRefBackpressureSink") {
+
+  val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+  require(maxBuffer > 0, "Buffer size must be greater than 0")
+
+  override def createLogicAndMaterializedValue: (GraphStageLogic, Unit) = {
+
+    val buffer: util.Deque[Out] = new util.ArrayDeque[Out]()
+    var acknowledgementReceived = false
+    var completeReceived = false
+    var pullRequestIsSent = true
```

Comment:
please put the variables inside the GraphStageLogic, doing it like this creates lots of 1-element arrays